components/browser/engine-gecko/src/main/java/mozilla/components/browser/engine/gecko/GeckoEngineSession.kt

```diff
@@ -56,6 +59,39 @@ class GeckoEngineSession(
         geckoSession.goForward()
     }
 
+    /**
+     * See [EngineSession.saveState]
+     *
+     * GeckoView provides a String representing the entire session state. We
+     * store this String using a single Map entry with key GECKO_STATE_KEY.
+
+     * See https://bugzilla.mozilla.org/show_bug.cgi?id=1441810 for
+     * discussion on sync vs. async, where a decision was made that
+     * callers should provide synchronous wrappers, if needed. In case we're
+     * asking for the state when persisting, a separate (independent) thread
+     * is used so we're not blocking anything else. In case of calling this
+     * method from onPause or similar, we also want a synchronous response.
+     */
+    override fun saveState(): Map<String, Any> = runBlocking {
+        val stateMap = CompletableDeferred<Map<String, Any>>()
+        geckoSession.saveState({ state ->
+            if (state != null) {
+                stateMap.complete(mapOf(GECKO_STATE_KEY to state.toString()))
+            } else {
+                stateMap.completeExceptionally(Throwable("Failed to save state"))
+            }
+        })
+        stateMap.await()
```

Comment:
As await() will rethrow the exception from above: Should we use something more specific than Throwable? Also: Should we make it more explicit that this method can throw an exception? E.g. [@Throws](https://kotlinlang.org/api/latest/jvm/stdlib/kotlin.jvm/-throws/index.html)